feature: add new `'u'` ("update") action for setqflist/setloclist

jfly opened the pull request:
This implements https://github.com/vim/vim/issues/15839

jfly committed:
feature: add `keep_idx` option to setqflist/setloclist

This is useful when replacing the quickfix list and you want to preserve
your current location in the list.

This implements https://github.com/vim/vim/issues/15839

jfly commented on src/testdir/test_quickfix.vim:
```diff
@@ -6370,7 +6370,7 @@ func Test_loclist_replace_autocmd()
   let s:bufnr = bufnr()
   cal setloclist(0, [{'0': 0, '': ''}])
-  au BufEnter * cal setloclist(1, [{'t': ''}, {'bufnr': s:bufnr}], 'r')
+  au BufEnter * cal setloclist(1, [{'t': ''}, {'bufnr': s:bufnr}, {'t': ''}], 'r')
```
This change is very non-obvious:

- This test replaced the loclist every time BufEnter occurs.
- Now that we preserve the loclist location when the loclist is being replaced, we *stay* on the selected item when selected it. Previously, we'd select the second item, which would switch buffers and trigger replacement of the loclist, which would then reset our selected loclist index to 1.
- Now, we stay on entry 2, and then we call `lnext`. This fails because there isn't a 3rd entry to navigate to.

The easiest fix to all this is just to add another entry to the loclist, so it's actually possible to navigate forwards in the loclist.

jfly replied:
Resolving. [We decided not to change the default behavior](https://github.com/vim/vim/pull/15841#pullrequestreview-2361372539), which means this test does not need to change.

chrisbra commented on src/quickfix.c:
```diff
@@ -12,6 +12,7 @@
  */
 
 #include "vim.h"
+#include "limits.h"
```
this should not be needed as it is in vim.h

jfly replied:
Ah, so it is! Removed.

jfly commented on runtime/doc/builtin.txt:
```diff
@@ -9787,6 +9787,8 @@ setqflist({list} [, {action} [, {what}]])		*setqflist()*
 		    items	list of quickfix entries. Same as the {list}
 				argument.
+		    keep_idx	select the entry nearest the currently
+		    		selected entry
```
I don't have much experience writing these. I fear this doesn't make a lot of sense. Feedback appreciated. Maybe I should say something like "useful when replacing the quickfix list and you want to keep your position in the list"?

jfly replied:
Marking this resolved. I think the updated docs for the new `'u'` action are ok. Feedback appreciated!

jfly reviewed (commented):
@chrisbra, I've hidden this new feature behind a new `keep_idx` option.

jfly commented on src/quickfix.c:
```diff
@@ -7822,7 +7914,19 @@ qf_set_properties(qf_info_T *qi, dict_T *what, int action, char_u *title)
+	if ((keep_idx_di = dict_find(what, (char_u *)"keep_idx", -1)) != NULL)
+	{
+	    if (keep_idx_di->di_tv.v_type != VAR_NUMBER)
+		return FAIL;
```
I'm not sure if it's OK to bail out out `qf_set_properties` here. The other code just sets `retval = FAIL` and then goes on to possibly do stuff after this block of `if` statements.

Feedback appreciated!

chrisbra replied:
Yeah,  I think following what has been done for the other properties is correct and should be done here as well.

jfly replied:
Ok, updated. This got kind of messy, I had to duplicate the call to `qf_setprop_items`.

I'm going to mark this resolved, but do let me know if you can think of a less bad way of doing this.

jfly replied:
For the record, this is irrelevant now that we're doing a `'u'` action instead of a `keep_idx` option.

jfly committed:
Preserve the "flow" of `qf_set_properties`

As with all the other codepaths, don't bail out immediately when
something go wrong. Instead, update `retval` and move on.

jfly committed:
Expand `Test_quickfix_keep_idx` to test for missing line number information

jfly committed:
Switch from `keep_idx` option to new `'u'` ("update") action

This cleaned the code up a bit =)

yegappan commented on src/testdir/test_quickfix.vim:
```diff
@@ -6462,6 +6462,32 @@ func Test_quickfix_buffer_contents()
 
+func Test_quickfix_update()
+  " Setup: populate a couple buffers
+  call setline(1, range(1,5))
```
You should create a new buffer before changing the contents of the buffer.

jfly replied:
Done!

yegappan commented on src/testdir/test_quickfix.vim:
```diff
@@ -6462,6 +6462,32 @@ func Test_quickfix_buffer_contents()
+  " Update the quickfix list again, but this time with missing line number
+  " information. Confirm that we keep the current buffer selected.
+  call setqflist([{'bufnr': b1}, {'bufnr': b2}], 'u')
+  call assert_equal(2, getqflist({'idx' : 0}).idx)
```
The test should clean up the newly created buffers.

jfly replied:
How do I ensure that happens even if the test fails?

jfly replied:
I've added a couple of `%bwipe!` commands at the end of the test, so I'm going to mark this done.

`<rant>`I skimmed the other tests in this file, and I'm not convinced they're being consistent about cleaning up the buffers they allocate. My inclination would be to try to solve this structurally rather than carefully: the test infrastructure/framework should make sure tests start with a clean slate, and furthermore, clean up any buffers/other stuff allocated during the test.`</rant>`

chrisbra replied:
I think one `%bwipe!` should be fine :)

jfly replied:
Oooh does that close all buffers? Ha! Got it.

Fixed.

jfly committed:
Allocate new buffers for this test, cleanup buffers at the end

jfly committed:
More complete code coverage

There are a lot of code paths through this `entry_is_closer_to_target`
function. I think we've got them all covered now.

jfly committed:
Oops: `%bwipe!` closes *all* buffers
